Objects/moduleobject.c

```diff
@@ -784,11 +786,73 @@ _PyModuleSpec_IsUninitializedSubmodule(PyObject *spec, PyObject *name)
     return rc;
 }
 
+static int
+_is_module_possibly_shadowing(PyObject *origin)
+{
+    // origin must be a unicode subtype
+    // Returns 1 if the module at origin could be shadowing a module of the
+    // same name later in the module search path. The condition we check is basically:
+    // not sys.flags.safe_path and os.path.dirname(origin) == (sys.path[0] or os.getcwd())
+    // Returns 0 otherwise (or if we aren't sure)
+    // Returns -1 if an error occurred that should be propagated
+    if (origin == NULL) {
+        return 0;
+    }
+    const PyConfig *config = _Py_GetConfig();
+    if (config->safe_path) {
+        return 0;
+    }
+
+    // os.path.dirname(origin)
+    wchar_t origin_dirname[MAXPATHLEN + 1];
+    Py_ssize_t size = PyUnicode_AsWideChar(origin, origin_dirname, MAXPATHLEN);
+    if (size < 0) {
+        return -1;
+    }
+    assert(size <= MAXPATHLEN);
+    origin_dirname[size] = L'\0';
+
+    wchar_t *sep = wcsrchr(origin_dirname, SEP);
+    if (!sep) {
+        return 0;
+    }
+    *sep = L'\0';
+
+    // sys.path[0] or os.getcwd()
+    wchar_t *sys_path_0 = config->sys_path_0;
+    if (!sys_path_0) {
+        if (!config->module_search_paths_set || config->module_search_paths.length == 0) {
+            return 0;
+        }
+        sys_path_0 = config->module_search_paths.items[0];
+        assert(sys_path_0 != NULL);
+    }
+    bool sys_path_allocated = false;
+    if (sys_path_0[0] == L'\0') {
+        // if sys.path[0] == "", treat it as if it were the current directory
+        sys_path_0 = PyMem_RawMalloc(MAXPATHLEN * sizeof(wchar_t));
+        if (!sys_path_0) {
+            return -1;
+        }
+        sys_path_allocated = true;
+        if (!_Py_wgetcwd(sys_path_0, MAXPATHLEN)) {
+            PyMem_RawFree(sys_path_0);
+            return -1;
+        }
+    }
+
+    int result = wcscmp(sys_path_0, origin_dirname) == 0;
```

Comment:
I was going to suggest making sure both filenames are absolute and fully resolved.  However, it's even more subtle:

* `python3 filename` - sys.path[0] is the file's directory
* `python3 -m name` - sys.path[0] is either CWD (before *almost* any user code runs) or "."
* `python3 -c ...` - sys.path[0] is the empty string
* `python3` (REPL) - sys.path[0] is the empty string

For those last two (and the "." case for -m), it uses CWD at the time the import happens.

Doesn't this check actually need to cover all those cases?

Reply:
looks like you mentioned this in https://github.com/python/cpython/pull/113769#discussion_r1475768452

Reply:
The current code should cover most of these cases, see the test cases here:
https://github.com/python/cpython/pull/113769/files#diff-ea25b7196846cae64e3ed2b3fb2e1ee420961442fcd801a07bcd9000178bbd58R804-R819

The "." case for `-m` seems pretty rare (need to not have realpath). It's easy enough to handle, but I'm not sure how to test that case.